cdisc_rules_engine/utilities/rule_processor.py

```diff
@@ -179,19 +179,23 @@ def rule_applies_to_class(self, rule, file_path, datasets: List[dict], domain: s
         if included_classes:
             if ALL_KEYWORD in included_classes:
                 return True
-            dataset = self.data_service.get_dataset(dataset_name=file_path)
+            variables = self.data_service.get_variables_metadata(
```

Comment:
This seems to be working well with pandas (33% runtime improvement), but failing with dask. The reason it fails with dask is these lines: https://github.com/cdisc-org/cdisc-rules-engine/blob/2b4bcefaf38866fcce0954f8f4b5464c10da81b3/cdisc_rules_engine/services/data_services/local_data_service.py#L168-L174

This is because the file is originally xpt, but when using dask, we convert the xpt to a parquet file for faster processing. So now when the file is read in, it is .parquet, which does not have an associated metadata reader.

I don't think the parquet files contain the original metadata, so I think you need to somehow pass the original filename to get the metadata, not the parquet filename. (The parquet filename should only be used for fetching the content.)

Reply:
when a dataset gets converted to parquet, the original file path is still in our dataservice.  I decided on just calling read_metadata again with the isolated original filepath instead of adding additional logic to process the filepath.  I dont believe the space or time complexity would be different and the readability is much nicer.  Let me know what you think. 
![image](https://github.com/cdisc-org/cdisc-rules-engine/assets/96841389/153f9114-0492-40cd-ad09-5773f7a4b6d0)

Reply:
I was hoping the file_path was already available, however, I don't think this code update is right.
For every call to read_metadata, it is now doing a read_metadata for every other file and it's not returning anything. We need to call read_metadat only for the current file_path's original file_path and we need to return the call to self.metadata.
We need a mapping or pointer from the current file's parquet file path to the xpt file path. Maybe this is already available somewhere, but if not, you might need to create it during the parquet file creation.

Reply:
@gerrycampion I made some fairly large edits.  When we get to https://github.com/cdisc-org/cdisc-rules-engine/blob/5d4b57a1a08fda57b6d9292cf98507d1df1d3f7f/scripts/run_validation.py#L154 the threads of the multiprocessing Pool unfortunately cannot access the state prior to the pool's creation so a simple map or pointer created before the pool would unfortunately not work.  I tried to use Multiprocessing's Manager to give access to a Datamap class to all the pool threads but, because the class had to be instantiated before the parquet file conversion, it would need to be a dependency and injected down to our data_service (I also had difficulty getting it to function properly).  I also considered redis but configuring and creating it seemed overkill for the utility it needed to provided.  I opted to make a json file we manipulate with the validator.  When I was forcing it to use dask, It appeared to be pulling metadata properly from the datasets.